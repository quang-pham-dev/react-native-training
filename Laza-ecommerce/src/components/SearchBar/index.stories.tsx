import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react-native';
import CenterView from '../../../storybook/stories/CenterView';
import SearchBar from './index';

storiesOf('Search Bar', module)
  .addDecorator(getStory => <CenterView>{getStory()}</CenterView>)
  .add('Default', () => <SearchBar onSubmitEditing={action('clicked')} />);
