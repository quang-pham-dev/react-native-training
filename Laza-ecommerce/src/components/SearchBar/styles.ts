import { StyleSheet } from 'react-native';

// Themes
import Colors from 'themes/Colors';
import Metrics from 'themes/Metrics';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    marginVertical: Metrics.margin.lg,
  },
  searchBarContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  inputWrapper: {
    flex: 1,
    flexDirection: 'row',
    height: 50,
    backgroundColor: Colors.secondaryBackground,
    borderRadius: Metrics.borderRadius.roundedMedium,
  },
  iconSearchWrapper: {
    width: Metrics.icons.standard,
    height: Metrics.icons.standard,
    alignSelf: 'center',
    justifyContent: 'center',
    marginLeft: Metrics.padding.xm,
  },
  iconSearch: {
    width: '100%',
    height: '100%',
  },
  input: {
    alignItems: 'flex-start',
    paddingLeft: Metrics.padding.md,
    paddingRight: 50,
  },
  iconVoiceWrapper: {
    marginLeft: Metrics.margin.sm,
  },
  iconVoice: {
    width: Metrics.icons.large,
    height: Metrics.icons.large,
    alignSelf: 'center',
  },
});

export default styles;
