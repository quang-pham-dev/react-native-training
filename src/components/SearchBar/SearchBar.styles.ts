import styled from 'styled-components/native'

// Styles
import FlexStyled from '@components/Flex/Flex.styles'

// Themes
import {Metrics, Colors} from '@themes'

export const SearchBarWrapperStyled = styled(FlexStyled.Row)`
  flex: 1;
  height: 50px;
  background-color: ${Colors.palette.lightGray};
  border-radius: ${Metrics.borderRadius.base}px;
`
export const IconWrapperStyled = styled.TouchableOpacity`
  justify-content: center;
  align-items: center;
  margin-left: ${Metrics.margin.small}px;
`

export const InputFiledStyled = styled.TextInput`
  flex: 1;
  padding-left: ${Metrics.padding.custom}px;
  padding-right: 20px;
  background-color: 'red';
`
