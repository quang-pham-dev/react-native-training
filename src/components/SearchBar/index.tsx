import React, {useCallback} from 'react'

// LIBS
import {s} from 'react-native-size-matters/extend'

// Styles
import IconStyled from '@components/Icon/Icon.styles'
import {Icons} from '@themes'
import {
  IconWrapperStyled,
  InputFiledStyled,
  SearchBarWrapperStyled,
} from './SearchBar.styles'
import FlexStyled from '@components/Flex/Flex.styles'

type SearchBarProps = {
  value?: string
  autoFocus?: boolean
  onSubmitEditing: () => void
  onChangeText?: (text: string) => void
}

const SearchBar = ({autoFocus, onSubmitEditing}: SearchBarProps) => {
  const [valueState, setValueState] = React.useState<string>('')

  const handleTextChange = useCallback((text: string) => {
    setValueState(text)
  }, [])

  const handleSubmitEditing = useCallback(() => {
    onSubmitEditing()
  }, [onSubmitEditing])

  return (
    <FlexStyled.Row>
      <SearchBarWrapperStyled>
        <>
          <IconWrapperStyled onPress={handleSubmitEditing}>
            <IconStyled width={s(20)} height={s(20)} source={Icons.search} />
          </IconWrapperStyled>
        </>
        <InputFiledStyled
          placeholder="Search.."
          autoFocus={autoFocus}
          autoCorrect={false}
          value={valueState}
          onChangeText={handleTextChange}
          onSubmitEditing={handleSubmitEditing}
        />
      </SearchBarWrapperStyled>
      <IconWrapperStyled>
        <IconStyled width={s(50)} height={s(50)} source={Icons.voice} />
      </IconWrapperStyled>
    </FlexStyled.Row>
  )
}

export default SearchBar
