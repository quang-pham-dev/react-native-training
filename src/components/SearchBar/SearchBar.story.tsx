import React from 'react'

// Storybook
import {action} from '@storybook/addon-actions'
import {storiesOf} from '@storybook/react-native'
import {StoryScreen, Story, UseCase} from '../../../storybook/views'

// Components
import SearchBar from './index'

declare let module: NodeModule

storiesOf('Search Bar', module)
  .addDecorator(fn => <StoryScreen>{fn()}</StoryScreen>)
  .add('Normal', () => (
    <Story>
      <UseCase text={'normal'}>
        <SearchBar onSubmitEditing={action('clicked')} />
      </UseCase>
    </Story>
  ))
