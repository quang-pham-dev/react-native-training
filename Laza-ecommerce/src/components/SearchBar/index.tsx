import React, { memo, useCallback, useContext } from 'react';
import { Image, TextInput, TouchableOpacity, View } from 'react-native';

// LIBS
import isEqual from 'react-fast-compare';

// Context
import { ProductsContext } from 'context/ProductsContext';
import { SEARCH_PRODUCTS_VALUE } from 'context/actions/products';

// Types
import { ISearchBarProps } from 'types/components/Search';

// Themes
import IMAGES from 'themes/Images';

// Styles
import styles from './styles';

const SearchBar = ({
  value,
  autoFocus,
  textInputStyles = {},
  onSubmitEditing,
}: ISearchBarProps) => {
  const { productDispatch } = useContext(ProductsContext);

  const [valueState, setValueState] = React.useState<string>('');

  const handleTextChange = useCallback(
    (text: string) => {
      setValueState(text);
      productDispatch({ type: SEARCH_PRODUCTS_VALUE, searchValue: text });
    },
    [setValueState],
  );

  const handleSubmitEditing = useCallback(() => {
    onSubmitEditing();
  }, [onSubmitEditing, valueState]);

  return (
    <View style={styles.container}>
      <View style={styles.searchBarContainer}>
        <View style={styles.inputWrapper}>
          <TouchableOpacity onPress={handleSubmitEditing} style={styles.iconSearchWrapper}>
            <Image style={styles.iconSearch} source={IMAGES.iconSearch} />
          </TouchableOpacity>
          <TextInput
            style={[styles.input, textInputStyles]}
            placeholder='Search..'
            autoFocus={autoFocus}
            autoCorrect={false}
            value={valueState}
            onChangeText={handleTextChange}
            onSubmitEditing={handleSubmitEditing}
          />
        </View>
        <TouchableOpacity style={styles.iconVoiceWrapper}>
          <Image style={styles.iconVoice} source={IMAGES.iconVoice} />
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default memo(SearchBar, isEqual);
