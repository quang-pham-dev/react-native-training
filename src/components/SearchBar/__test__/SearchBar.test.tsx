import React from 'react'

// LIBS
import {render} from '@testing-library/react-native'

// Components
import SearchBar from '@components/SearchBar'

describe('Search Bar Component', () => {
  const props = {
    value: 'search value',
    onSubmitEditing: jest.fn(),
  }
  const tree = render(<SearchBar {...props} />)
  test('Should render correctly', () => {
    const component = tree.toJSON()
    expect(component).toMatchSnapshot()
  })
})
