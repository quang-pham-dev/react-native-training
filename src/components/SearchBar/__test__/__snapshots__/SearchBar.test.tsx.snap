// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar Component Should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#F5F6FA",
          "borderRadius": 10,
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 50,
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginLeft": 10,
          "opacity": 1,
        }
      }
    >
      <Image
        height={20}
        source={
          Object {
            "testUri": "../../../src/assets/icons/search.webp",
          }
        }
        style={
          Array [
            Object {},
          ]
        }
        width={20}
      />
    </View>
    <TextInput
      autoCorrect={false}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Search.."
      style={
        Array [
          Object {
            "backgroundColor": "'red'",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "paddingLeft": 10,
            "paddingRight": 20,
          },
        ]
      }
      value=""
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginLeft": 10,
        "opacity": 1,
      }
    }
  >
    <Image
      height={50}
      source={
        Object {
          "testUri": "../../../src/assets/icons/voice.webp",
        }
      }
      style={
        Array [
          Object {},
        ]
      }
      width={50}
    />
  </View>
</View>
`;
